src/sundials/stl/sunstl_vector.h

```diff
@@ -0,0 +1,173 @@
+}
+
+static inline sunbooleantype MAKE_NAME(SUNStlVectorTtype,
+                                       IsEmpty)(SUNStlVectorTtype self)
+{
+  return self->size == 0;
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype, Resize)(SUNStlVectorTtype self,
+                                                        size_t new_capacity)
+{
+  if (new_capacity <= self->capacity) return;
+  TTYPE* new_values = (TTYPE*)realloc(self->values, sizeof(TTYPE) * new_capacity);
+  self->values   = new_values;
+  self->capacity = new_capacity;
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype, Grow)(SUNStlVectorTtype self)
+{
+  if (self->size == self->capacity)
+  {
+    size_t new_capacity = (size_t)(ceil(((double)self->capacity) * GROWTH_FACTOR));
+    MAKE_NAME(SUNStlVectorTtype, Resize)(self, new_capacity);
+  }
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype,
+                             PushBack)(SUNStlVectorTtype self, TTYPE element)
+{
+  if (self->size == self->capacity)
+  {
+    MAKE_NAME(SUNStlVectorTtype, Grow)(self);
+  }
+  self->values[self->size++] = element;
+}
+
+static inline TTYPE* MAKE_NAME(SUNStlVectorTtype, At)(SUNStlVectorTtype self,
+                                                      size_t index)
+{
+  if (index >= self->size)
+  {
+    // Handle index out of bounds
+    return NULL;
+  }
+  return &(self->values[index]);
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype, Set)(SUNStlVectorTtype self,
+                                                     size_t index, TTYPE element)
+{
+  if (index >= self->size)
+  {
+    // Handle index out of bounds
+    return;
+  }
+  self->values[index] = element;
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype, PopBack)(SUNStlVectorTtype self)
+{
+  static TTYPE nullish;
+  if (self->size == 0) return;
+  self->size--;
+  MAKE_NAME(SUNStlVectorTtype, Set)(self, self->size, nullish);
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype, Erase)(SUNStlVectorTtype self,
+                                                       size_t index)
+{
+  static TTYPE nullish;
+  if (self->size == 0) return;
+  MAKE_NAME(SUNStlVectorTtype, Set)(self, index, nullish);
+  for (size_t i = index; i < self->size - 1; i++)
+  {
+    self->values[i]     = self->values[i + 1];
+    self->values[i + 1] = nullish;
+  }
+  self->size -= 1;
+}
+
+static inline size_t MAKE_NAME(SUNStlVectorTtype, Size)(SUNStlVectorTtype self)
+{
+  return self->size;
+}
+
+static inline size_t MAKE_NAME(SUNStlVectorTtype, Capacity)(SUNStlVectorTtype self)
+{
+  return self->capacity;
+}
+
+static inline void MAKE_NAME(SUNStlVectorTtype,
+                             Destroy)(SUNStlVectorTtype* self_ptr)
+{
+  static TTYPE nullish;
+
+  if (!self_ptr || !(*self_ptr)) return;
+
+  SUNStlVectorTtype self = *self_ptr;
+
+  for (size_t i = 0; i < MAKE_NAME(SUNStlVectorTtype, Size)(self); i++)
+  {
+    self->destroyValue(&(self->values[i]));
+    self->values[i] = nullish;
+  }
+
+  *self_ptr = NULL;
+}
+
+#undef TTYPE
```

Comment:
No. The `#undef` allows for you to include this header multiple times to have multiple instantiations of the vector with different types.

Reply:
Sure, but the "includer" could do that just as easily, allowing both the `#define` and `#undef` to be co-located, instead of implicitly expecting the `#define` to be cleaned up elsewhere.  I could imagine that an "includer" who only wants one instantiation could potentially reuse `TTYPE` later in their code.